Orange/statistics/util.py

```diff
@@ -46,44 +50,72 @@ def bincount(X, max_val=None, weights=None, minlength=None):
             nans = (~nonnan).sum(axis=0)
     else:
         nans = 0. if X.ndim == 1 else np.zeros(X.shape[1], dtype=float)
+
     if minlength is None and max_val is not None:
         minlength = max_val + 1
-    bc = np.array([]) if minlength is not None and minlength <= 0 else \
-        np.bincount(X.astype(np.int32, copy=False),
-                    weights=weights, minlength=minlength).astype(float)
+
+    if minlength is not None and minlength <= 0:
+        bc = np.array([])
+    else:
+        bc = np.bincount(
+            X.astype(np.int32, copy=False), weights=weights, minlength=minlength
+        ).astype(float)
+        # Since `csr_matrix.values` only contain non-zero values, we must count
+        # those separately and set the appropriate bin
+        if sp.issparse(X_):
+            bc[0] = np.prod(X_.shape) - X_.nnz
+
     return bc, nans
 
 
 def countnans(X, weights=None, axis=None, dtype=None, keepdims=False):
     """
-    Count the undefined elements in arr along given axis.
+    Count the undefined elements in an array along given axis.
 
     Parameters
     ----------
     X : array_like
     weights : array_like
         Weights to weight the nans with, before or after counting (depending
         on the weights shape).
+    axis : Optional[int]
 
     Returns
     -------
-    counts
+    Union[np.ndarray, float]
+
     """
     if not sp.issparse(X):
         X = np.asanyarray(X)
         isnan = np.isnan(X)
         if weights is not None and weights.shape == X.shape:
             isnan = isnan * weights
+
+        # In order to keep return types consistent with sparse vectors, we will
+        # handle `axis=1` given a regular 1d numpy array equivallently as
```

Comment:
I agree, this does seem to be the most sensible solution. Most of all, we should be consistent with the behaviour on sparse and dense matrices, otherwise much confusion is inevitable.

Reply:
> Most of all, we should be consistent with the behaviour on sparse and dense matrices, otherwise much confusion is inevitable.

But on the other hand, changing user-provided input of `axis=1` to `axis=0` just because `axis=1` would result in an error is IMO worse than some slight incompatibility between sparse and dense. If I could come up with at least one use-case where this incompatibility would really hurt us I might be less against it.

IMO if you want to stick to 100% same behavior, I would prefer if we also raise an error when a sparse row is provided and user passes axis=1. Hence we could essentially treat a sparse matrix of shape (1, X) as 1D array and get exactly the same behaviour as with 1D dense matrices.

Reply:
Yes, I agree, changing the user input was a bad idea. The error is much better.